From f5d29555b53457e8172552dbb0228bf178bbd4a8 Mon Sep 17 00:00:00 2001
From: Siddharth Ray <siddharthr@google.com>
Date: Sun, 17 Jun 2018 15:02:38 -0700
Subject: [PATCH 05/70] Fix in Modem Activity Info updates

A check is added to capture the following error conditions:
- Idle time is larger than the total activity duration.
- Sleep time is larger than the total activity duration.
- Rx time is larger than the total activity duration.
- Total Tx time is larger than the total activity duration.

BUG:110319204
Change-Id: I47b18313c0d0e336110859bd79da2347862bbfde
Merged-In: I47b18313c0d0e336110859bd79da2347862bbfde
---
 .../android/phone/PhoneInterfaceManager.java  | 21 ++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/src/com/android/phone/PhoneInterfaceManager.java b/src/com/android/phone/PhoneInterfaceManager.java
index ae3ec4eab..101ab44e5 100755
--- a/src/com/android/phone/PhoneInterfaceManager.java
+++ b/src/com/android/phone/PhoneInterfaceManager.java
@@ -3683,7 +3683,7 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
         synchronized (mLastModemActivityInfo) {
             ModemActivityInfo info = (ModemActivityInfo) sendRequest(CMD_GET_MODEM_ACTIVITY_INFO,
                     null);
-            if (info != null) {
+            if (isModemActivityInfoValid(info)) {
                 int[] mergedTxTimeMs = new int[ModemActivityInfo.TX_POWER_LEVELS];
                 for (int i = 0; i < mergedTxTimeMs.length; i++) {
                     mergedTxTimeMs[i] =
@@ -3712,6 +3712,25 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
         result.send(0, bundle);
     }
 
+    // Checks that ModemActivityInfo is valid. Sleep time, Idle time, Rx time and Tx time should be
+    // less than total activity duration.
+    private boolean isModemActivityInfoValid(ModemActivityInfo info) {
+        if (info == null) {
+            return false;
+        }
+        int activityDurationMs =
+            (int) (info.getTimestamp() - mLastModemActivityInfo.getTimestamp());
+        int totalTxTimeMs = 0;
+        for (int i = 0; i < info.getTxTimeMillis().length; i++) {
+            totalTxTimeMs += info.getTxTimeMillis()[i];
+        }
+        return (info.isValid()
+            && (info.getSleepTimeMillis() <= activityDurationMs)
+            && (info.getIdleTimeMillis() <= activityDurationMs)
+            && (info.getRxTimeMillis() <= activityDurationMs)
+            && (totalTxTimeMs <= activityDurationMs));
+    }
+
     /**
      * {@hide}
      * Returns the service state information on specified subscription.
-- 
2.17.1

